### PR TITLE
Stay on Git branch when using `--since` with `query changes`

### LINKED ISSUE #2423: `plz query changes --since` leaves Git repo in detached HEAD state
This is a minor annoyance:

```
csn@debian-11:~/plztest$ git log
commit 7c895251323334ce2c500a10d4b8354f900a0f06 (HEAD, tag: v0.0.2, tag: v0.0.1, new-branch)
Author: Chris Novakovic <chris@chrisn.me.uk>
Date:   Sun Apr 10 22:58:10 2022 +0100

    Modify README

commit 71562e561f015c4b5e9d39f78f6ec12cecda5cbe (master)
Author: Chris Novakovic <chris@chrisn.me.uk>
Date:   Sun Apr 10 22:57:18 2022 +0100

    Add README.md
```

(`v0.0.1` is a lightweight tag, `v0.0.2` is an annotated tag.)

```
csn@debian-11:~/plztest$ git status
On branch new-branch
nothing to commit, working tree clean
csn@debian-11:~/plztest$ plz query changes --since master
csn@debian-11:~/plztest$ git status
HEAD detached at 7c89525
nothing to commit, working tree clean
```

One way to fix this would be to use the output of `git symbolic-ref -q --short HEAD` as the revision to switch back to after checking out the revision passed via `--since`, falling back to `git rev-parse HEAD` if `git symbolic-ref` exits unsuccessfully (i.e., if the working tree isn't on a branch). I'm not sure of the best way to do that, though - the output of `scm.CurrentRevIdentifier` and `scm.DescribeIdentifier` are both used as inputs to the stamp generation function, and making `CurrentRevIdentifier` output a branch name means they could both potentially return branch or tag names, which might cause stale outputs if tags or branches are shifted around:

```
csn@debian-11:~/plztest$ git status
On branch new-branch
nothing to commit, working tree clean

csn@debian-11:~/plztest$ git symbolic-ref -q --short HEAD
new-branch
# CurrentRevIdentifier:
csn@debian-11:~/plztest$ git rev-parse HEAD
7c895251323334ce2c500a10d4b8354f900a0f06
# DescribeIdentifier:
csn@debian-11:~/plztest$ git describe --always 7c895251323334ce2c500a10d4b8354f900a0f06
v0.0.2
```

It's a bit ugly, but perhaps a new function `scm.CurrentRevReference` that does what I described above, and is used exclusively by `plz query changes`? So we'd have:

* `CurrentRevIdentifier`: returns the full commit hash of whatever `HEAD` is pointing to
* `CurrentRevReference`: returns a reference to whatever `HEAD` is pointing to
* `DescribeIdentifier`: returns a human-readable reference to whatever `HEAD` is pointing to (could be a local or remote branch name, lightweight tag name, or short hash)

### PR DESCRIPTION
**Pull request by chrisnovakovic**
When used on a Git repository, `plz query changes --since <S>` gets the commit hash of the current HEAD `C`, checks out the revision `S`, then checks out `C`. If the repository was on a branch before the command was run, Please restores the working tree to `C` rather than the branch for which `C` is the HEAD, leaving the repository in detached HEAD state.

Before checking out `S`, run `git symbolic-ref -q --short HEAD` to determine the name of the current branch (if any), falling back to `git rev-parse HEAD` if this fails (which indicates that the working tree is not on a branch), and use this as the value of `C`. This ensures the repository is always restored to the state it was in before `plz query changes` was executed.

Closes #2423.

### CODE REVIEW

**Comment by chrisnovakovic** [comment]
Here's a demo of how `plz query changes` behaves now.

```
(0) csn@debian-11:~/plztest$ git log
commit 7c895251323334ce2c500a10d4b8354f900a0f06 (HEAD, tag: v0.0.2, tag: v0.0.1, new-branch)
Author: Chris Novakovic <chris@chrisn.me.uk>
Date:   Sun Apr 10 22:58:10 2022 +0100

    Modify README

commit 71562e561f015c4b5e9d39f78f6ec12cecda5cbe (master)
Author: Chris Novakovic <chris@chrisn.me.uk>
Date:   Sun Apr 10 22:57:18 2022 +0100

    Add README.md
```

When on a branch:

```
(0) csn@debian-11:~/plztest$ git checkout new-branch
Switched to branch 'new-branch'
(0) csn@debian-11:~/plztest$ git status
On branch new-branch
nothing to commit, working tree clean
(0) csn@debian-11:~/plztest$ plz query changes --since master
(0) csn@debian-11:~/plztest$ git status
On branch new-branch
nothing to commit, working tree clean
```

When in detached HEAD state:

```
(0) csn@debian-11:~/plztest$ git checkout 7c895251323334ce2c500a10d4b8354f900a0f06
HEAD is now at 7c89525 Modify README
(0) csn@debian-11:~/plztest$ git status
HEAD detached at 7c89525
nothing to commit, working tree clean
(0) csn@debian-11:~/plztest$ plz query changes --since master
(0) csn@debian-11:~/plztest$ git status
HEAD detached at 7c89525
nothing to commit, working tree clean
```

Checking out a tag is equivalent to checking out a commit, which leaves the repo in detached HEAD state (as above):

```
(0) csn@debian-11:~/plztest$ git checkout v0.0.1
HEAD is now at 7c89525 Modify README
(0) csn@debian-11:~/plztest$ git status
HEAD detached at 7c89525
nothing to commit, working tree clean
(0) csn@debian-11:~/plztest$ plz query changes --since master
(0) csn@debian-11:~/plztest$ git status
HEAD detached at 7c89525
nothing to commit, working tree clean
```

**Review by tiagovtristao — APPROVED** [review]
Nice, looks good to me, thanks. Let's get another approval first, before landing this, just in case.

**Review by peterebden — APPROVED** [review]
Nice, thanks! This has always been a little irritating to me but never found the right git command, no matter how long I searched through https://git-man-page-generator.lokaltog.net ;-)

**Pull request merged**
